dashboard/database.py

```diff
@@ -0,0 +1,140 @@
+    """Returns the N most sold tracks in the database."""
+
+    print("Collating most popular tracks.")
+
+    query = """
+        SELECT T.title, A.name, COUNT(*) AS copies_sold
+        FROM track_purchase AS PT
+        JOIN track AS T
+        USING(track_id)
+        JOIN artist as A
+        USING(artist_id)
+        GROUP BY T.title, A.name
+        ORDER BY copies_sold DESC
+        LIMIT %s
+        ;
+        """
+
+    with _conn.cursor() as cur:
+        cur.execute(query, (n,))
+        data = cur.fetchall()
+
+    return pd.DataFrame(data)
+
+
+@st.cache_resource(ttl="1hr")
+def get_popular_albums(_conn: Connection, n: int = 5) -> pd.DataFrame:
+    """Returns the N most sold albums in the database."""
+
+    print("Collating most popular albums.")
+
+    query = """
+        SELECT AB.title, AT.name, COUNT(*) AS copies_sold
+        FROM album_purchase AS PA
+        JOIN album AS AB
+        USING(album_id)
+        JOIN artist as AT
+        USING(artist_id)
+        GROUP BY AB.title, AT.name
+        ORDER BY copies_sold DESC
+        LIMIT %s
+        ;
+        """
+
+    with _conn.cursor() as cur:
+        cur.execute(query, (n,))
+        data = cur.fetchall()
+
+    return pd.DataFrame(data)
+
+
+@st.cache_resource(ttl="1hr")
+def get_popular_artists(_conn: Connection, n: int = 5) -> pd.DataFrame:
+    """Returns the N artists with the most sales in the database."""
+
+    print("Collating most popular artists.")
+
+    query = """
+        SELECT A.name, COUNT(AP.album_purchase_id)+COUNT(TP.track_purchase_id) AS total_sales
+        FROM artist as A
+        JOIN album AS AB
+        USING(artist_id)
+        JOIN album_purchase AS AP
+        USING(album_id)
+        JOIN track as T
+        USING(artist_id)
+        JOIN track_purchase as TP
+        USING(track_id)
+        GROUP BY AB.title, A.name
+        ORDER BY total_sales DESC
+        LIMIT %s
+        ;
+        """
+
+    with _conn.cursor() as cur:
+        cur.execute(query, (n,))
+        data = cur.fetchall()
+
+    return pd.DataFrame(data)
+
+
+@st.cache_resource(ttl="1hr")
+def get_sales_by_tag(_conn: Connection) -> pd.DataFrame:
+    """Returns a count of sales by genre/tag."""
+
+    print("Counting sales by tag.")
+
+    query = """
+        SELECT TG.name AS tag, COUNT(*) AS sales
+        FROM tag AS TG
+        LEFT JOIN album_tag_assignment AS ATG
+        USING(tag_id)
+        LEFT JOIN track_tag_assignment AS TGA
+        USING(tag_id)
+        LEFT JOIN album AS A
+        USING(album_id)
+        LEFT JOIN track AS T
+        USING(track_id)
+        LEFT JOIN album_purchase AS AP
+        USING(album_id)
+        LEFT JOIN track_purchase AS TP
+        USING(track_id)
+        GROUP BY tag
+        ;
+        """
+
+    with _conn.cursor() as cur:
+        cur.execute(query)
+        data = cur.fetchall()
+
+    return pd.DataFrame(data)
+
+
+if __name__ == "__main__":
+    load_dotenv()
```

Comment:
Have you closed your connection?

Reply:
Ah good spot. I don't need the main here because I'm going to call the functions in this file elsewhere so I'll just get rid of it altogether.